grammar.js

```diff
@@ -110,17 +135,51 @@ module.exports = grammar({
       $.template_body
     ),
 
+    // The EBNF makes a distinction between function type parameters and other
+    // type parameters as you can't specify variance on function type
+    // parameters. This isn't important to the structure of the AST so we don't
+    // make that distinction.
     type_parameters: $ => seq(
       '[',
-      commaSep1($._type_parameter),
+      commaSep1($._variant_type_parameter),
       ']'
     ),
 
-    _type_parameter: $ => choice(
-      '_',
-      $.identifier
+    _variant_type_parameter: $ => seq(
+      choice(
+        $.covariant_type_parameter,
+        $.contravariant_type_parameter,
+        $._type_parameter // invariant type parameter
+      )
+    ),
+
+    covariant_type_parameter: $ => seq(
+      '+',
+      $._type_parameter
     ),
 
+    contravariant_type_parameter: $ => seq(
+      '-',
+      $._type_parameter,
+    ),
+
+    _type_parameter: $ => seq(
+      choice($.wildcard, $.identifier),
+      optional($.type_parameters),
+      optional($.upper_bound),
+      optional($.lower_bound),
+      optional(repeat($.view_bound)),
+      optional(repeat($.context_bound)),
+    ),
```

Comment:
Now that I'm seeing how much structure there is to type parameters, I think we should tweak their structure so that there is always a visible syntax node corresponding to each parameter (as opposed to the `identifier` and `upper_bound` node appearing as siblings within the `type_parameters` node).

Maybe something like this:

```
==================
Type parameters
==================

class A[B, +C <: D, -E : F : G]

---

(compilation_unit
  (class_definition
    (identifier)
    (type_parameters
      (invariant_type_parameter (type_identifier))
      (covariant_type_parameter
        (type_identifier)
        (upper_bound (type_identifier)))
      (contravariant_type_parameter
        (type_identifier)
        (context_bound (type_identifier))
        (context_bound (type_identifier))))))
```

Does that make sense? We could have a hidden helper rule called `_type_parameter` that is shared between the three. In addition, we should add the `_type_parameter` to the `inline` section of the grammar. That will avoid actually creating that node at runtime for performance reasons.